core/src/main/java/com/predic8/membrane/core/interceptor/rest/REST2SOAPInterceptor.java

```diff
@@ -183,21 +185,25 @@ public Outcome handleResponse(Exchange exc) throws Exception {
 		header.setContentType(TEXT_XML_UTF8);
 
 		XML2HTTP.unwrapMessageIfNecessary(exc.getResponse());
-		convertResponseToJSONIfNecessary(exc.getRequest().getHeader(), mapping, exc.getResponse(), exc.getStringProperties());
+		convertResponseToJSONIfNecessary(exc.getRequest().getHeader(), exc.getResponse(), exc.getStringProperties());
 
-		return Outcome.CONTINUE;
+		return CONTINUE;
 	}
 
+	private void convertResponseToJSONIfNecessary(Header header, Response response, Map<String, String> properties) throws Exception {
+		if (!response.isXML())
+			return;
 
-	private static final MediaType[] supportedTypes = convertStringsToMediaType(new String[] { TEXT_XML, APPLICATION_JSON_UTF8 });
+		// Not JSON
+		if (!getMediaTypeWithHighestQualityFromAcceptHeader(header).getSubtype().contains("json"))
+			return;
 
-	private void convertResponseToJSONIfNecessary(Header requestHeader, Mapping mapping, Response response, Map<String, String> properties) throws Exception {
-		boolean inputIsXml = response.isXML();
-		int wantedType = requestHeader.getBestAcceptedType(supportedTypes);
-		if (inputIsXml && wantedType >= 1) {
-			response.setBodyContent(xml2json(response.getBodyAsStreamDecoded(), properties));
-			setJSONContentType(response.getHeader());
-		}
+		response.setBodyContent(xml2json(response.getBodyAsStreamDecoded(), properties));
+		setJSONContentType(response.getHeader());
+	}
+
+	private static MediaType getMediaTypeWithHighestQualityFromAcceptHeader(Header requestHeader) {
+		return sortMimeTypeByQualityFactorAscending(requestHeader.getFirstValue(ACCEPT)).get(0);
```

Comment:
possible IndexOutOfBoundsException, if the list is empty

Reply:
Fixed